androidapp/Crisislab/app/src/main/kotlin/com/example/crisislab/SocketListener.kt

```diff
@@ -0,0 +1,146 @@
+package com.example.crisislab
+
+import NotificationHandler
+import android.content.Intent
+import android.os.Build
+import android.util.Log
+import androidx.annotation.RequiresApi
+import okhttp3.Response
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import org.json.JSONObject
+import java.time.*
+import java.time.format.DateTimeFormatter
+import kotlin.math.round
+
+// Listens for WebSocket events and processes messages accordingly
+class SocketListener(
+    logViewModel: LogViewModel,
+    socketStatusViewModel: SocketStatusViewModel,
+    context: MainActivity
+) : WebSocketListener() {
+
+    // ViewModel to manage logs
+    var logViewModel: LogViewModel = logViewModel
+    // ViewModel to manage socket status
+    var socketStatusViewModel: SocketStatusViewModel = socketStatusViewModel
+    // Context of the main activity
+    var context: MainActivity = context
+
+    // Called when the WebSocket connection is opened
+    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
+    override fun onOpen(webSocket: WebSocket, response: Response) {
+        // Initialize the notification handler
+        context.notificationHandler = NotificationHandler(
+            context.notificationModule.provideNotificationManager(context),
+            context
+        )
+
+        // Start the notification service if it's not running
+        if(!context.notificationHandler.isServiceRunning) {
+            output("Service not running")
+            val intent = Intent(context, context.notificationHandler::class.java)
+            context.startService(intent)
+            context.notificationHandler.isServiceRunning = true
+            output("Notification Service began.")
+        }
+
+        // Update socket status to connected
+        socketStatusViewModel.updateStatus("Status: Connected.")
+    }
+
+    // Called when a message is received from the WebSocket
+    @RequiresApi(Build.VERSION_CODES.TIRAMISU)
+    override fun onMessage(webSocket: WebSocket, text: String) {
+        val data = JSONObject(text)
+        val prevAlerts = data.optJSONArray("previous_alerts")
+
+        // Process previous alerts if any
+        if(prevAlerts != null) {
+            for(i in 0 until prevAlerts.length()) {
+                val prevPacket: JSONObject = prevAlerts.getJSONObject(i)
+                val currentPacket = checkPacket(prevPacket)
+
+                if(currentPacket.bool) {
+                    logPacket(currentPacket.packet, false)
+                }
+            }
+        }
+
+        // Check and log the current packet
+        val checked = checkPacket(data)
+        if (checked.bool) {
+            logPacket(checked.packet, true)
+        }
+    }
+
+    // Data class to hold packet information
+    data class packetType(val bool: Boolean, val packet: HashMap<String, String?>)
```

Comment:
a) I feel like `packetType` could just be `Packet` (`PacketType` feels like it should be an enum)
b) what is `bool` supposed to represent?

Reply:
a) Fixed
b) Need to check what exactly this does with Max lol